src/ripple/conditions/RsaSha256.h

```diff
@@ -0,0 +1,302 @@
+
+        return RSA_verify_PKCS1_PSS(key, digest.data(), EVP_sha256(), buf.data(), -1) == 1;
+    }
+
+public:
+    RsaSha256 () = default;
+
+    Condition
+    condition() const override
+    {
+        std::vector<std::uint8_t> m;
+        m.reserve (1024);
+
+        oer::encode_octetstring (
+            modulus_.size(),
+            modulus_.begin(),
+            modulus_.end(),
+            std::back_inserter(m));
+
+        sha256_hasher h;
+        h (m.data(), m.size());
+
+        Condition cc;
+        cc.type = type();
+        cc.featureBitmask = features();
+        cc.maxFulfillmentLength = payloadSize();
+
+        cc.fingerprint = static_cast<sha256_hasher::result_type>(h);
+
+        return cc;
+    }
+
+    std::uint16_t
+    type () const override
+    {
+        return condition_rsa_sha256;
+    }
+
+    std::uint32_t
+    features () const override
+    {
+        return feature_rsa_pss | feature_sha256;
+    }
+
+    std::size_t
+    payloadSize () const override
+    {
+        return
+            oer::predict_octetstring_size (modulus_.size()) +
+            oer::predict_octetstring_size (signature_.size());
+    }
+
+    Buffer
+    payload() const override
+    {
+        Buffer b (payloadSize());
+
+        auto out = oer::encode_octetstring (
+            modulus_.size(),
+            modulus_.begin(),
+            modulus_.end(),
+            b.begin());
+
+        oer::encode_octetstring (
+            signature_.size(),
+            signature_.begin(),
+            signature_.end(),
+            out);
+
+        return b;
+    }
+
+    bool
+    validate (Slice data) const override
+    {
+        RsaKey rsa (RSA_new());
+
+        rsa->n = BN_new();
+        BN_bin2bn(modulus_.data(), modulus_.size(), rsa->n);
+
+        rsa->e = BN_new();
+        BN_set_word (rsa->e, 65537);
+
+        return validate_helper (rsa.get(), data);
+    }
+
+    /** Sign the given message with an RSA key */
+    bool sign (
+        std::string const& key,
+        Slice message)
+    {
+        // This ugly const_cast/reinterpret_cast is needed
+        // on some machines. Although the documentation
+        // suggests that the function accepts a void const*
+        // argument, apparently some platforms have OpenSSL
+        // libraries that are up-to-date but accept void*.
+        auto bio = BIO_new_mem_buf(
+            const_cast<void*>(static_cast<void const*>(key.data())),
+            key.size());
+
+        if (!bio)
+            return false;
+
+        RsaKey rsa (PEM_read_bio_RSAPrivateKey(
+            bio, NULL, NULL, NULL));
+
+        BIO_free(bio);
+
+        if (!rsa)
+            return false;
+
+        return sign_helper (rsa.get(), message);
+    }
+
+    bool
+    parse_payload (Slice s) override
+    {
+        auto iter = s.begin ();
+
+        std::size_t len;
+
+        std::tie (iter, len) = oer::decode_length (
+            iter, s.end());
+
+        std::memcpy (modulus_.alloc (len), iter, len);
```

Comment:
`len` is untrused. We need to verify that we aren't reading beyond the end of the slice or someone can create a payload that will crash us.